README.rst
==========
ProcRunner
==========


.. image:: https://img.shields.io/pypi/v/procrunner.svg
        :target: https://pypi.python.org/pypi/procrunner
        :alt: PyPI release

.. image:: https://img.shields.io/conda/vn/conda-forge/procrunner.svg
        :target: https://anaconda.org/conda-forge/procrunner
        :alt: Conda Version

.. image:: https://dev.azure.com/DLS-tooling/procrunner/_apis/build/status/CI?branchName=master
        :target: https://github.com/DiamondLightSource/python-procrunner/commits/master
        :alt: Build status

.. image:: https://ci.appveyor.com/api/projects/status/jtq4brwri5q18d0u/branch/master
        :target: https://ci.appveyor.com/project/Anthchirp/python-procrunner
        :alt: Build status

.. image:: https://readthedocs.org/projects/procrunner/badge/?version=latest
        :target: https://procrunner.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://img.shields.io/pypi/pyversions/procrunner.svg
        :target: https://pypi.python.org/pypi/procrunner
        :alt: Supported Python versions

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
        :target: https://github.com/ambv/black
        :alt: Code style: black

Versatile utility function to run external processes

* Free software: BSD license
* Documentation: https://procrunner.readthedocs.io.


Features
--------

* runs an external process and waits for it to finish
* does not deadlock, no matter the process stdout/stderr output behaviour
* returns the exit code, stdout, stderr (separately, both as bytestrings),
  as a subprocess.CompletedProcess object
* process can run in a custom environment, either as a modification of
  the current environment or in a new environment from scratch
* stdin can be fed to the process
* stdout and stderr is printed by default, can be disabled
* stdout and stderr can be passed to any arbitrary function for
  live processing (separately, both as unicode strings)
* optionally enforces a time limit on the process, raising a
  subprocess.TimeoutExpired exception if it is exceeded.


Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
History
==================

3.0.0 (2022-01-??)
------------------
* Drop Python 3.6, 3.7 support
* The run() function now returns a subprocess.CompletedProcess object,
  which no longer allows array access operations
  (those were deprecated in `#60 <https://github.com/DiamondLightSource/python-procrunner/pull/60>`_)
* The run() argument 'raise_timeout_exception' is now set by default,
  a 'False' value will lead to a UserWarning and a behavioural change.
  The argument is now deprecated and will be removed in a future version.
  (previously introduced in `#61 <https://github.com/DiamondLightSource/python-procrunner/pull/61>`_)
* Calling the run() function with multiple unnamed arguments is no longer supported
  (previously deprecated in `#62 <https://github.com/DiamondLightSource/python-procrunner/pull/62>`_)
* The run() function no longer accepts a 'debug' argument
  (previously deprecated in `#63 <https://github.com/DiamondLightSource/python-procrunner/pull/63>`_)

2.3.3 (2022-03-23)
------------------
* Allow specifying 'preexec_fn' and 'creationflags' keywords, which will be passed through to
  the subprocess call

2.3.2 (2022-01-28)
------------------
* The run() function now understands stdin=subprocess.DEVNULL to close the subprocess stdin,
  rather than to connect through the existing stdin, which is the current default

2.3.1 (2021-10-25)
------------------
* Add Python 3.10 support

2.3.0 (2020-10-29)
------------------
* Add Python 3.9 support, drop Python 3.5 support
* Fix a file descriptor leak on subprocess execution

2.2.0 (2020-09-07)
------------------
* Calling the run() function with unnamed arguments (other than the command
  list as the first argument) is now deprecated. As a number of arguments
  will be removed in a future version the use of unnamed arguments will
  cause future confusion. `Use explicit keyword arguments instead (#62). <https://github.com/DiamondLightSource/python-procrunner/pull/62>`_
* `The run() function debug argument has been deprecated (#63). <https://github.com/DiamondLightSource/python-procrunner/pull/63>`_
  This is only used to debug the NonBlockingStream* classes. Those are due
  to be replaced in a future release, so the argument will no longer serve
  a purpose. Debugging information remains available via standard logging
  mechanisms.
* Final version supporting Python 3.5

2.1.0 (2020-09-05)
------------------
* `Deprecated array access on the return object (#60). <https://github.com/DiamondLightSource/python-procrunner/pull/60>`_
  The return object will become a subprocess.CompletedProcess in a future
  release, which no longer allows array-based access. For a translation table
  of array elements to attributes please see the pull request linked above.
* Add a `new parameter 'raise_timeout_exception' (#61). <https://github.com/DiamondLightSource/python-procrunner/pull/61>`_
  When set to 'True' a subprocess.TimeoutExpired exception is raised when the
  process runtime exceeds the timeout threshold. This defaults to 'False' and
  will be set to 'True' in a future release.

2.0.0 (2020-06-24)
------------------
* Python 3.5+ only, support for Python 2.7 has been dropped
* Deprecated function alias run_process() has been removed
* Fixed a stability issue on Windows

1.1.0 (2019-11-04)
------------------
* Add Python 3.8 support, drop Python 3.4 support

1.0.2 (2019-05-20)
------------------
* Stop environment override variables leaking into the process environment

1.0.1 (2019-04-16)
------------------
* Minor fixes on the return object (implement equality,
  mark as unhashable)

1.0.0 (2019-03-25)
------------------
* Support file system path objects (PEP-519) in arguments
* Change the return object to make it similar to
  subprocess.CompletedProcess, introduced with Python 3.5+

0.9.1 (2019-02-22)
------------------
* Have deprecation warnings point to correct code locations

0.9.0 (2018-12-07)
------------------
* Trap UnicodeEncodeError when printing output. Offending characters
  are replaced and a warning is logged once. Hints at incorrectly set
  PYTHONIOENCODING.

0.8.1 (2018-12-04)
------------------
* Fix a few deprecation warnings

0.8.0 (2018-10-09)
------------------
* Add parameter working_directory to set the working directory
  of the subprocess

0.7.2 (2018-10-05)
------------------
* Officially support Python 3.7

0.7.1 (2018-09-03)
------------------
* Accept environment variable overriding with numeric values.

0.7.0 (2018-05-13)
------------------
* Unicode fixes. Fix crash on invalid UTF-8 input.
* Clarify that stdout/stderr values are returned as bytestrings.
* Callbacks receive the data decoded as UTF-8 unicode strings
  with unknown characters replaced by \ufffd (unicode replacement
  character). Same applies to printing of output.
* Mark stdin broken on Windows.

0.6.1 (2018-05-02)
------------------
* Maintenance release to add some tests for executable resolution.

0.6.0 (2018-05-02)
------------------
* Fix Win32 API executable resolution for commands containing a dot ('.') in
  addition to a file extension (say '.bat').

0.5.1 (2018-04-27)
------------------
* Fix Win32API dependency installation on Windows.

0.5.0 (2018-04-26)
------------------
* New keyword 'win32resolve' which only takes effect on Windows and is enabled
  by default. This causes procrunner to call the Win32 API FindExecutable()
  function to try and lookup non-.exe files with the corresponding name. This

tests/test_procrunner_system.py
from __future__ import annotations

import os
import subprocess
import sys
import timeit

import pytest

import procrunner


def test_simple_command_invocation():
    if os.name == "nt":
        command = ["cmd.exe", "/c", "echo", "hello"]
    else:
        command = ["echo", "hello"]

    result = procrunner.run(command)

    assert result.returncode == 0
    assert result.stdout == b"hello" + os.linesep.encode("utf-8")
    assert result.stderr == b""


def test_simple_command_invocation_with_closed_stdin():
    if os.name == "nt":
        command = ["cmd.exe", "/c", "echo", "hello"]
    else:
        command = ["echo", "hello"]

    result = procrunner.run(command, stdin=subprocess.DEVNULL)

    assert result.returncode == 0
    assert result.stdout == b"hello" + os.linesep.encode("utf-8")
    assert result.stderr == b""


def test_decode_invalid_utf8_input(capsys):
    test_string = b"test\xa0string\n"
    if os.name == "nt":
        pytest.xfail("Test requires stdin feature which does not work on Windows")
        command = ["cmd.exe", "/c", "type", "CON"]
    else:
        command = ["cat"]
    result = procrunner.run(command, stdin=test_string)
    assert result.returncode == 0
    assert not result.stderr
    if os.name == "nt":
        # Windows modifies line endings
        assert result.stdout == test_string[:-1] + b"\r\n"
    else:
        assert result.stdout == test_string
    out, err = capsys.readouterr()
    assert out == "test\ufffdstring\n"
    assert err == ""


def test_running_wget(tmp_path):
    command = ["wget", "https://www.google.com", "-O", "-"]
    try:
        result = procrunner.run(command, working_directory=tmp_path)
    except OSError as e:
        if e.errno == 2:
            pytest.skip("wget not available")
        raise
    assert result.returncode == 0
    assert b"http" in result.stderr
    assert b"google" in result.stdout


def test_path_object_resolution(tmp_path):
    sentinel_value = b"sentinel"
    tmp_path.joinpath("tempfile").write_bytes(sentinel_value)
    tmp_path.joinpath("reader.py").write_text(
        "with open('tempfile') as fh:\n    print(fh.read())"
    )
    assert "LEAK_DETECTOR" not in os.environ
    result = procrunner.run(
        [sys.executable, tmp_path / "reader.py"],
        environment_override={"PYTHONHASHSEED": "random", "LEAK_DETECTOR": "1"},
        working_directory=tmp_path,
    )
    assert result.returncode == 0
    assert not result.stderr
    assert sentinel_value == result.stdout.strip()
    assert (
        "LEAK_DETECTOR" not in os.environ
    ), "overridden environment variable leaked into parent process"


def test_timeout_behaviour_old_legacy(tmp_path):
    command = (sys.executable, "-c", "import time; time.sleep(5)")
    start = timeit.default_timer()
    try:
        with pytest.raises(subprocess.TimeoutExpired) as te:
            with pytest.warns(UserWarning, match="timeout"):
                procrunner.run(
                    command,
                    timeout=0.1,
                    working_directory=tmp_path,
                    raise_timeout_exception=False,
                )
    except RuntimeError:
        # This test sometimes fails with a RuntimeError.
        runtime = timeit.default_timer() - start
        assert runtime < 3
        return
    runtime = timeit.default_timer() - start
    assert runtime < 3
    assert te.value.stdout == b""
    assert te.value.stderr == b""
    assert te.value.timeout == 0.1
    assert te.value.cmd == command


def test_timeout_behaviour_legacy(tmp_path):
    command = (sys.executable, "-c", "import time; time.sleep(5)")
    start = timeit.default_timer()
    try:
        with pytest.raises(subprocess.TimeoutExpired) as te:
            with pytest.warns(DeprecationWarning, match="timeout"):
                procrunner.run(
                    command,
                    timeout=0.1,
                    working_directory=tmp_path,
                    raise_timeout_exception=True,
                )
    except RuntimeError:
        # This test sometimes fails with a RuntimeError.
        runtime = timeit.default_timer() - start
        assert runtime < 3
        return
    runtime = timeit.default_timer() - start
    assert runtime < 3
    assert te.value.stdout == b""
    assert te.value.stderr == b""
    assert te.value.timeout == 0.1
    assert te.value.cmd == command


def test_timeout_behaviour(tmp_path):
    command = (sys.executable, "-c", "import time; time.sleep(5)")
    start = timeit.default_timer()
    try:
        with pytest.raises(subprocess.TimeoutExpired) as te:
            procrunner.run(
                command,
                timeout=0.1,
                working_directory=tmp_path,
            )
    except RuntimeError:
        # This test sometimes fails with a RuntimeError.
        runtime = timeit.default_timer() - start
        assert runtime < 3
        return
    runtime = timeit.default_timer() - start
    assert runtime < 3
    assert te.value.stdout == b""
    assert te.value.stderr == b""
    assert te.value.timeout == 0.1
    assert te.value.cmd == command


tests/test_procrunner_resolution.py
import os
import sys

import pytest

import procrunner


def PEP519(path):
    class MockObject:
        @staticmethod
        def __fspath__():
            return path

        def __repr__(self):
            return "<path object: %s>" % path

    return MockObject()


@pytest.mark.parametrize(
    "obj",
    (
        None,
        True,
        False,
        1,
        1.0,
        ["thing"],
        {},
        {1},
        {"thing": "thing"},
        "string",
        b"bytes",
        RuntimeError(),
        ["thing", PEP519("thing")],  # no recursive resolution
    ),
)
def test_path_object_resolution_for_non_path_objs_does_not_modify_objects(obj):
    assert procrunner._path_resolve(obj) is obj


def test_path_object_resolution_of_path_objects():
    assert procrunner._path_resolve(PEP519("thing")) == "thing"


@pytest.mark.skipif(sys.platform != "win32", reason="windows specific test only")
def test_name_resolution_for_simple_exe():
    command = ["cmd.exe", "/c", "echo", "hello"]

    resolved = procrunner._windows_resolve(command)

    # command should be replaced with full path to cmd.exe
    assert resolved[0].lower().endswith("\\cmd.exe")
    assert os.path.exists(resolved[0])

    # parameters are unchanged
    assert resolved[1:] == tuple(command[1:])


@pytest.mark.skipif(sys.platform != "win32", reason="windows specific test only")
def test_name_resolution_for_complex_cases(tmp_path):
    bat = "simple_bat_extension"
    cmd = "simple_cmd_extension"
    exe = "simple_exe_extension"
    dotshort = "more_complex_filename_with_a.dot"
    dotlong = "more_complex_filename.withadot"

    (tmp_path / (bat + ".bat")).touch()
    (tmp_path / (cmd + ".cmd")).touch()
    (tmp_path / (exe + ".exe")).touch()
    (tmp_path / (dotshort + ".bat")).touch()
    (tmp_path / (dotlong + ".cmd")).touch()

    def is_valid(command):
        assert len(command) == 1
        assert os.path.exists(tmp_path / command[0])

    is_valid(procrunner._windows_resolve([bat], path=os.fspath(tmp_path)))
    is_valid(procrunner._windows_resolve([cmd], path=os.fspath(tmp_path)))
    is_valid(procrunner._windows_resolve([exe], path=os.fspath(tmp_path)))
    is_valid(procrunner._windows_resolve([dotshort], path=os.fspath(tmp_path)))
    is_valid(procrunner._windows_resolve([dotlong], path=os.fspath(tmp_path)))


src/procrunner/__init__.py
from __future__ import annotations

import codecs
import io
import logging
import os
import select
import shutil
import subprocess
import sys
import time
import timeit
import warnings
from multiprocessing import Pipe
from threading import Thread
from typing import Any, Callable, Optional, Union

#
#  run() - A function to synchronously run an external process, supporting
#          the following features:
#
#    - runs an external process and waits for it to finish
#    - does not deadlock, no matter the process stdout/stderr output behaviour
#    - returns the exit code, stdout, stderr (separately) as a
#      subprocess.CompletedProcess object
#    - process can run in a custom environment, either as a modification of
#      the current environment or in a new environment from scratch
#    - stdin can be fed to the process
#    - stdout and stderr is printed by default, can be disabled
#    - stdout and stderr can be passed to any arbitrary function for
#      live processing
#    - optionally enforces a time limit on the process
#
#
#  Usage example:
#
# import procrunner
# result = procrunner.run(['/bin/ls', '/some/path/containing spaces'])
#
#  Returns:
#
# subprocess.CompletedProcess(
#   args=('/bin/ls', '/some/path/containing spaces'),
#   returncode=2,
#   stdout=b'',
#   stderr=b'/bin/ls: cannot access /some/path/containing spaces: No such file or directory\n'
# )

__version__ = "2.3.3"

logger = logging.getLogger("procrunner")
logger.addHandler(logging.NullHandler())


class _LineAggregator:
    """
    Buffer that can be filled with stream data and will aggregate complete
    lines. Lines can be printed or passed to an arbitrary callback function.
    The lines passed to the callback function are UTF-8 decoded and do not
    contain a trailing newline character.
    """

    def __init__(self, print_line=False, callback=None):
        """Create aggregator object."""
        self._buffer = ""
        self._print = print_line
        self._callback = callback
        self._decoder = codecs.getincrementaldecoder("utf-8")("replace")

    def add(self, data):
        """
        Add a single character to buffer. If one or more full lines are found,
        print them (if desired) and pass to callback function.
        """
        data = self._decoder.decode(data)
        if not data:
            return
        self._buffer += data
        if "\n" in data:
            to_print, remainder = self._buffer.rsplit("\n")
            if self._print:
                try:
                    print(to_print)
                except UnicodeEncodeError:
                    print(to_print.encode(sys.getdefaultencoding(), errors="replace"))
                    if not hasattr(self, "_warned"):
                        logger.warning("output encoding error, characters replaced")
                        setattr(self, "_warned", True)
            if self._callback:
                self._callback(to_print)
            self._buffer = remainder

    def flush(self):
        """Print/send any remaining data to callback function."""
        self._buffer += self._decoder.decode(b"", final=True)
        if self._buffer:
            if self._print:
                print(self._buffer)
            if self._callback:
                self._callback(self._buffer)
        self._buffer = ""


class _NonBlockingStreamReader:
    """Reads a stream in a thread to avoid blocking/deadlocks"""

    def __init__(self, stream, output=True, debug=False, notify=None, callback=None):
        """Creates and starts a thread which reads from a stream."""
        self._buffer = io.BytesIO()
        self._closed = False
        self._closing = False
        self._debug = debug
        self._stream = stream
        self._terminated = False

        def _thread_write_stream_to_buffer():
            la = _LineAggregator(print_line=output, callback=callback)
            char = True
            while char:
                if select.select([self._stream], [], [], 0.1)[0]:
                    char = self._stream.read(1)
                    if char:
                        self._buffer.write(char)
                        la.add(char)
                else:
                    if self._closing:
                        break
            self._stream.close()
            self._terminated = True
            la.flush()
            if self._debug:
                logger.debug("Stream reader terminated")
            if notify:
                notify()

        def _thread_write_stream_to_buffer_windows():
            line = True
            while line:
                line = self._stream.readline()
                if line:
                    self._buffer.write(line)
                    if output or callback:
                        linedecode = line.decode("utf-8", "replace")
                        if output:
                            print(linedecode)
                        if callback:
                            callback(linedecode)
            self._stream.close()
            self._terminated = True
            if self._debug:
                logger.debug("Stream reader terminated")
            if notify:
                notify()

        if os.name == "nt":
            self._thread = Thread(target=_thread_write_stream_to_buffer_windows)
        else:
            self._thread = Thread(target=_thread_write_stream_to_buffer)
        self._thread.daemon = True
        self._thread.start()

    def has_finished(self):
        """
        Returns whether the thread reading from the stream is still alive.
        """
        return self._terminated

    def get_output(self):
        """
        Retrieve the stored data in full.
        This call may block if the reading thread has not yet terminated.
        """
        self._closing = True
        if not self.has_finished():
            if self._debug:
                # Main thread overtook stream reading thread.
                underrun_debug_timer = timeit.default_timer()
                logger.warning("NBSR underrun")
            self._thread.join()
            if not self.has_finished():
                if self._debug:
                    logger.debug(
                        "NBSR join after %f seconds, underrun not resolved",
                        timeit.default_timer() - underrun_debug_timer,
                    )
                raise Exception("thread did not terminate")
            if self._debug:
                logger.debug(
                    "NBSR underrun resolved after %f seconds",
                    timeit.default_timer() - underrun_debug_timer,
                )
        if self._closed:
            raise Exception("streamreader double-closed")
        self._closed = True
        data = self._buffer.getvalue()
        self._buffer.close()
        return data


class _NonBlockingStreamWriter:
    """Writes to a stream in a thread to avoid blocking/deadlocks"""

    def __init__(self, stream, data, debug=False, notify=None):
        """Creates and starts a thread which writes data to stream."""
        self._buffer = data
        self._buffer_len = len(data)
        self._buffer_pos = 0
        self._max_block_len = 4096
        self._stream = stream
        self._terminated = False

        def _thread_write_buffer_to_stream():
            while self._buffer_pos < self._buffer_len:
                if (self._buffer_len - self._buffer_pos) > self._max_block_len:
                    block = self._buffer[
                        self._buffer_pos : (self._buffer_pos + self._max_block_len)
                    ]
                else:
                    block = self._buffer[self._buffer_pos :]
                try:
                    self._stream.write(block)
                except OSError as e:
                    if (
                        e.errno == 32
                    ):  # broken pipe, ie. process terminated without reading entire stdin
                        self._stream.close()
                        self._terminated = True
                        if notify:
                            notify()
                        return
                    raise
                self._buffer_pos += len(block)
                if debug:
                    logger.debug("wrote %d bytes to stream", len(block))
            self._stream.close()
            self._terminated = True
            if notify:
                notify()

        self._thread = Thread(target=_thread_write_buffer_to_stream)
        self._thread.daemon = True
        self._thread.start()

    def has_finished(self):
        """Returns whether the thread writing to the stream is still alive."""
        return self._terminated

    def bytes_sent(self):
        """Return the number of bytes written so far."""
        return self._buffer_pos

    def bytes_remaining(self):
        """Return the number of bytes still to be written."""
        return self._buffer_len - self._buffer_pos


def _path_resolve(obj):
    """
    Resolve file system path (PEP-519) objects to strings.

    :param obj: A file system path object or something else.
    :return: A string representation of a file system path object or, for
             anything that was not a file system path object, the original
             object.
    """
    if obj and hasattr(obj, "__fspath__"):
        return obj.__fspath__()
    return obj


def _windows_resolve(command, path=None):
    """
    Try and find the full path and file extension of the executable to run.
    This is so that e.g. calls to 'somescript' will point at 'somescript.cmd'
    without the need to set shell=True in the subprocess.

    :param command: The command array to be run, with the first element being
                    the command with or w/o path, with or w/o extension.
    :return: Returns the command array with the executable resolved with the
             correct extension. If the executable cannot be resolved for any
             reason the original command array is returned.
    """
    if not command or not isinstance(command[0], str):
        return command

    found_executable = shutil.which(command[0], path=path)
    if found_executable:
        logger.debug("Resolved %s as %s", command[0], found_executable)
        return (found_executable, *command[1:])

    if "\\" in command[0]:
        # Special case. shutil.which may not detect file extensions if a full
        # path is given, so try to resolve the executable explicitly
        for extension in os.getenv("PATHEXT").split(os.pathsep):
            found_executable = shutil.which(command[0] + extension, path=path)
            if found_executable:
                return (found_executable, *command[1:])

    logger.warning("Error trying to resolve the executable: %s", command[0])
    return command


def run(
    command,
    *,
    timeout: Optional[float] = None,
    callback_stderr: Optional[Callable] = None,
    callback_stdout: Optional[Callable] = None,
    creationflags: int = 0,
    environment: Optional[dict[str, str]] = None,
    environment_override: Optional[dict[str, str]] = None,
    preexec_fn: Optional[Callable] = None,
    print_stderr: bool = True,
    print_stdout: bool = True,
    raise_timeout_exception: Any = ...,
    stdin: Optional[Union[bytes, int]] = None,
    win32resolve: bool = True,
    working_directory: Optional[str] = None,
) -> subprocess.CompletedProcess:
    """
    Run an external process.

    File system path objects (PEP-519) are accepted in the command, environment,
    and working directory arguments.

    :param array command: Command line to be run, specified as array.
    :param timeout: Terminate program execution after this many seconds.
    :param stdin: Optional bytestring that is passed to command stdin,
                  or subprocess.DEVNULL to disable stdin.
    :param boolean print_stdout: Pass stdout through to sys.stdout.
    :param boolean print_stderr: Pass stderr through to sys.stderr.
    :param callback_stdout: Optional function which is called for each
                            stdout line.
    :param callback_stderr: Optional function which is called for each
                            stderr line.
    :param creationflags: flags that will be passed to subprocess call
    :param dict environment: The full execution environment for the command.
    :param dict environment_override: Change environment variables from the
                                      current values for command execution.
    :param preexec_fn: pre-execution function, will be passed to subprocess call
    :param boolean win32resolve: If on Windows, find the appropriate executable
                                 first. This allows running of .bat, .cmd, etc.
                                 files without explicitly specifying their
                                 extension.
    :param string working_directory: If specified, run the executable from
                                     within this working directory.
    :param boolean raise_timeout_exception: Deprecated compatibility flag.
    :return: The exit code, stdout, stderr (separately, as byte strings)
             as a subprocess.CompletedProcess object.
    """

    logger.debug("Starting external process: %s", command)

    if stdin is None:
        stdin_pipe = None
    elif isinstance(stdin, int):
        assert (
            stdin == subprocess.DEVNULL
        ), "stdin argument only allows subprocess.DEVNULL as numeric argument"
        stdin_pipe = subprocess.DEVNULL
        stdin = None
    else:
        assert sys.platform != "win32", "stdin argument not supported on Windows"
        stdin_pipe = subprocess.PIPE

    start_time = timeit.default_timer()
    if timeout is not None:
        max_time = start_time + timeout
    if not raise_timeout_exception:
        warnings.warn(
            "Using procrunner with raise_timeout_exception=False is no longer supported",
            UserWarning,
            stacklevel=3,
        )
    elif raise_timeout_exception is True:
        warnings.warn(
            "The raise_timeout_exception argument is deprecated and will be removed in a future release",
            DeprecationWarning,
            stacklevel=3,
        )

    if environment is not None:
        env = {key: _path_resolve(environment[key]) for key in environment}
    else:
        env = {key: value for key, value in os.environ.items()}
    if environment_override:
        env.update(
            {
                key: str(_path_resolve(environment_override[key]))
                for key in environment_override
            }
        )

    command = tuple(_path_resolve(part) for part in command)
    if win32resolve and sys.platform == "win32":
        command = _windows_resolve(command)

    p = subprocess.Popen(
        command,
        shell=False,
        cwd=working_directory,
        env=env,
        stdin=stdin_pipe,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        creationflags=creationflags,
        preexec_fn=preexec_fn,
    )

    thread_pipe_pool = []
    notifyee, notifier = Pipe(False)
    thread_pipe_pool.append(notifyee)
    stdout = _NonBlockingStreamReader(
        p.stdout,
        output=print_stdout,
        notify=notifier.close,
        callback=callback_stdout,
    )
    notifyee, notifier = Pipe(False)
    thread_pipe_pool.append(notifyee)
    stderr = _NonBlockingStreamReader(
        p.stderr,
        output=print_stderr,
        notify=notifier.close,
        callback=callback_stderr,
    )
    if stdin is not None:
        notifyee, notifier = Pipe(False)
        thread_pipe_pool.append(notifyee)
        _NonBlockingStreamWriter(p.stdin, data=stdin, notify=notifier.close)

    timeout_encountered = False

    while (p.returncode is None) and (
        (timeout is None) or (timeit.default_timer() < max_time)
    ):
        # wait for some time or until a stream is closed
        try:
            if thread_pipe_pool:
                # Wait for up to 0.5 seconds or for a signal on a remaining stream,
                # which could indicate that the process has terminated.
                try:
                    event = thread_pipe_pool[0].poll(0.5)
                except BrokenPipeError as e:
                    # on Windows this raises "BrokenPipeError: [Errno 109] The pipe has been ended"
                    # which is for all intents and purposes equivalent to a True return value.
                    if e.winerror != 109:
                        raise
                    event = True
                if event:
                    # One-shot, so remove stream and watch remaining streams
                    thread_pipe_pool.pop(0)
            else:
                time.sleep(0.5)
        except KeyboardInterrupt:
            p.kill()  # if user pressed Ctrl+C we won't be able to produce a proper report anyway
            # but at least make sure the child process dies with us
            raise

        # check if process is still running
        p.poll()

    if p.returncode is None:
        # timeout condition
        timeout_encountered = True
        logger.debug("timeout (T%.2fs)", timeit.default_timer() - max_time)

        # send terminate signal and wait some time for buffers to be read
        p.terminate()
        if thread_pipe_pool:
            try:
                thread_pipe_pool[0].poll(0.5)
            except BrokenPipeError as e:
                # on Windows this raises "BrokenPipeError: [Errno 109] The pipe has been ended"
                # which is for all intents and purposes equivalent to a True return value.
                if e.winerror != 109:
                    raise
                thread_pipe_pool.pop(0)
        if not stdout.has_finished() or not stderr.has_finished():
            time.sleep(2)
        p.poll()

    if p.returncode is None:
        # thread still alive
        # send kill signal and wait some more time for buffers to be read
        p.kill()
        if thread_pipe_pool:
            try:
                thread_pipe_pool[0].poll(0.5)
            except BrokenPipeError as e:
                # on Windows this raises "BrokenPipeError: [Errno 109] The pipe has been ended"
                # which is for all intents and purposes equivalent to a True return value.
                if e.winerror != 109:
                    raise
                thread_pipe_pool.pop(0)
        if not stdout.has_finished() or not stderr.has_finished():
            time.sleep(5)
        p.poll()

    if p.returncode is None:
        raise RuntimeError("Process won't terminate")

    runtime = timeit.default_timer() - start_time
    if timeout is not None:
        logger.debug(
            "Process ended after %.1f seconds with exit code %d (T%.2fs)",
            runtime,
            p.returncode,
            timeit.default_timer() - max_time,
        )
    else:
        logger.debug(
            "Process ended after %.1f seconds with exit code %d", runtime, p.returncode
        )

    output_stdout = stdout.get_output()
    output_stderr = stderr.get_output()

    if timeout is not None and timeout_encountered:
        raise subprocess.TimeoutExpired(
            cmd=command, timeout=timeout, output=output_stdout, stderr=output_stderr
        )

    return subprocess.CompletedProcess(
        args=command,
        returncode=p.returncode,
        stdout=output_stdout,
        stderr=output_stderr,
    )
